### PR TITLE
improve user scribbles algorithm - fixes 266 and 262

### LINKED ISSUE #262: Scribbles - Fix installation requirements for scribbles app
Currently it's complicated.. gpu vs no-gpu and use requirements.txt to add any app specific packages

### PR DESCRIPTION
**Pull request by masadcv**
This pull request fixes #262  and #266 .

It updates/adds the following fixes to user scribbles algorithm:
- Updates make unary algorithm to match with what is in the BIFSeg paper
- Updated to save, load and use network logits (pre-softmax) in user scribbles processing algorithm
- Added a softening function to soften resulting probabilities from logits to be within the range [0.1, 0.9]
- Updated scribble->scribbles throughout the code/ui to be more consistent and to fix a bug in ui that prevented selection of different scribbles method in drop down slicer panel
- Updated lines in Restored transform, so that a logits with dimension N, H, W, D can be written to disk
- To address issue #262 , moved relevant package installation into requirements.txt and instructing use of `BUILD_MONAI=1 pip install -r requirements.txt` within README.md for relevant scribbles app

### CODE REVIEW

File: monailabel/utils/others/post.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ def __call__(self, data):
                 resizer = Resize(spatial_size=spatial_size, mode=self.mode[idx])
                 result = resizer(result, mode=self.mode[idx], align_corners=self.align_corners[idx])
 
-            d[key] = result if len(result.shape) <= 3 else result[0]
+            d[key] = result if len(result.shape) <= 3 else result[0] if result.shape[0] == 1 else result
```
Comment (masadcv) [inline monailabel/utils/others/post.py:99]: @SachidanandAlle I had to add this line in order to allow logits with dimension N, H, W, D to be written back to data dictionary. Maybe there is a better way or perhaps I missed something here? As this is common code with other methods so better to check if we are not breaking something here. Could you have a look please?

File: sample-apps/segmentation_spleen_postproc/README.md [hunk 2]
```diff
@@ -12,32 +11,24 @@ The flow of this application is outlined below:
...
-# Installing pre-requisites
+# Using the app
 
-## 1. Install MONAI with BUILD_MONAI=1
-Uses CRF layer from MONAI, which requires compiling the C++/CUDA code following instructions from [MONAI docs](https://docs.monai.io/en/latest/installation.html#option-1-as-a-part-of-your-system-wide-module). 
-This can be done by uninstalling any previous monai/monai-weekly version and running the following command:
+## Install relevant requirements
 
-`BUILD_MONAI=1 pip install git+https://github.com/Project-MONAI/MONAI#egg=monai`
+`BUILD_MONAI=1 pip install -r requirements.txt`
```
Comment (masadcv) [inline sample-apps/segmentation_spleen_postproc/README.md:18]: @SachidanandAlle @aihsani this is the proposed way of installing requirements through requirements.txt. I have simplified to the README to include this detail and remove the pre-requisite section. I couldnt find a way to bypass having to specify `BUILD_MONAI=1` for now. Maybe there is a way?

File: monailabel/utils/others/post.py (hunk 1 above)
Comment (SachidanandAlle) [reply monailabel/utils/others/post.py:99]: best way to check/verify is run some deepgrow 2D/3D actions in slicer

File: sample-apps/segmentation_spleen_postproc/README.md (hunk 2 above)
Comment (SachidanandAlle) [reply sample-apps/segmentation_spleen_postproc/README.md:18]: lets think about this... how to pass additional ENV for the app..

File: sample-apps/segmentation_spleen_postproc/lib/scribbles.py [hunk 3]
```diff
@@ -192,33 +203,33 @@ class SpleenBIFSegSimpleCRF(SpleenPostProc):
...
-        return [
-            LoadImaged(keys=["image", "logits", "label"]),
-            AddChanneld(keys=["image", "logits", "label"]),
-            # at the moment Simple CRF implementation is bottleneck taking a long time,
-            # therefore scaling non-isotropic with big spacing
-            Spacingd(keys=["image", "logits"], pixdim=[3.5, 3.5, 5.0]),
-            Spacingd(keys=["label"], pixdim=[3.5, 3.5, 5.0], mode="nearest"),
-            ScaleIntensityRanged(keys="image", a_min=-300, a_max=200, b_min=0.0, b_max=1.0, clip=True),
-        ]
+    # def pre_transforms(self):
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_postproc/lib/scribbles.py:210]: any big chunk of unused code.. better to remove it

File: sample-apps/segmentation_spleen_postproc/lib/transforms.py [hunk 4]
```diff
@@ -588,18 +622,30 @@ def __call__(self, data):
...
+        self.ref_image = ref_image if ref_image else key
         self.result = result
+        self.meta_key_postfix = meta_key_postfix
 
     def __call__(self, data):
+        logger = logging.getLogger(self.__class__.__name__)
+
         d = dict(data)
-        writer = Writer(label=self.key)
+        file_ext = "".join(pathlib.Path(d["image_path"]).suffixes)
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_postproc/lib/transforms.py:635]: you can still use MONAI writer with nibabel=True.. and avoid writing 10 lines of code here.. 
`writer = Writer(label=self.key, nibabel=True)`

**Comment by SachidanandAlle** [comment]
Next time, its better to split the fixes into 2 PRs.. 
first to fix the bugs kind of issues.. second one to improve the methods

**Comment by SachidanandAlle** [comment]
```
/home/runner/work/MONAILabel/MONAILabel/sample-apps/segmentation_spleen_postproc/lib/transforms.py:13:1: F401 
'monai.transforms.MapTransform' imported but unused

1     F401 'monai.transforms.MapTransform' imported but unused
```

**Comment by SachidanandAlle** [comment]
Also good to rename segmentation_spleen_postproc => segmentation_spleen_scribbles

File: sample-apps/segmentation_spleen_postproc/lib/transforms.py (hunk 4 above)
Comment (masadcv) [reply sample-apps/segmentation_spleen_postproc/lib/transforms.py:635]: Thanks! I will update this.

File: sample-apps/segmentation_spleen_postproc/lib/scribbles.py (hunk 3 above)
Comment (masadcv) [reply sample-apps/segmentation_spleen_postproc/lib/scribbles.py:210]: Sure, I will remove this and any other similar chunks.

File: sample-apps/segmentation_spleen_scribbles/main.py [hunk 5]
```diff
@@ -33,10 +33,10 @@ def init_infers(self):
...
                 self.final_model, load_from_mmar(self.mmar, self.model_dir)
             ),
-            "BIFSeg+CRF": SpleenBIFSegCRF(),
-            "BIFSeg+SimpleCRF": SpleenBIFSegSimpleCRF(),
-            "BIFSeg+GraphCut": SpleenBIFSegGraphCut(),
-            "Int.+BIFSeg+GraphCut": SpleenInteractiveGraphCut(),
+            "ISeg+CRF": SpleenISegCRF(),
+            "ISeg+SimpleCRF": SpleenISegSimpleCRF(),
+            "ISeg+GraphCut": SpleenISegGraphCut(),
+            "Int.+ISeg+GraphCut": SpleenInteractiveGraphCut(),
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_scribbles/main.py:39]: `Int.+ISeg+GraphCut` => There is a dot..  
`Int.` doesn't sound normal..

File: sample-apps/segmentation_spleen_scribbles/main.py [hunk 6]
```diff
@@ -33,10 +33,10 @@ def init_infers(self):
             "Spleen_Segmentation": SegmentationWithWriteLogits(
                 self.final_model, load_from_mmar(self.mmar, self.model_dir)
             ),
-            "BIFSeg+CRF": SpleenBIFSegCRF(),
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_scribbles/main.py:36]: Can we define/describe all these methods in README.md for the app.

File: sample-apps/segmentation_spleen_scribbles/README.md [hunk 7]
```diff
@@ -12,32 +11,24 @@ The flow of this application is outlined below:
...
-
-`pip install simplecrf`
-
-# Running the app
-
-## Server
+### Server
 On the server side, run server app using the following command:
 
 `CUDA_VISIBLE_DEVICES=0 ./monailabel start_server --app ../sample-apps/segmentation_spleen_postproc/ --studies /path/to/dataset/Task09_Spleen/imagesTrSmall/`
```
Comment (SachidanandAlle) [inline sample-apps/segmentation_spleen_scribbles/README.md:29]: can we change it to `monailabel start_server --app .... `

users must have already installed the monailabel from pip

**Comment by SachidanandAlle** [comment]
- Info.yaml still has name postprocs..
- When I submit the label.. I can't use scribbles anymore.. tool is enabled but something is cleared and I can't paint anymore for foreground/background.. 
- Don't add ISeg+CRF as it doesn't work due to MONAI Core issue (which is still not solved..) please create an issue and tag it to monai core issue for tracking
- Since we disable ISeg+CRF you can remove BUILD_MONAI dependency from requirements.txt. 
- Can we make ISeg+GraphCut as first item (reorder in init) which is better over ISeg+CRF

Also in the readme better to add a note that it tested for Ubunutu and may work for windows (depends on simplecrf)

File: monailabel/utils/others/post.py (hunk 1 above)
Comment (SachidanandAlle) [reply monailabel/utils/others/post.py:99]: verified

File: sample-apps/segmentation_spleen_scribbles/main.py (hunk 5 above)
Comment (masadcv) [reply sample-apps/segmentation_spleen_scribbles/main.py]: Fixed

**Comment by masadcv** [comment]
> * Info.yaml still has name postprocs..
This has been fixed now. 

> * When I submit the label.. I can't use scribbles anymore.. tool is enabled but something is cleared and I can't paint anymore for foreground/background..
> * Don't add ISeg+CRF as it doesn't work due to MONAI Core issue (which is still not solved..) please create an issue and tag it to monai core issue for tracking
> * Since we disable ISeg+CRF you can remove BUILD_MONAI dependency from requirements.txt.
> * Can we make ISeg+GraphCut as first item (reorder in init) which is better over ISeg+CRF

The bug is in GPU implementation of CRF. I have switched to use CPU implementation by default for ISeg+CRF. I will add a separate issue to document this and maybe reopen the issue on MONAI core related to this. As a side effect of using CPU, this transform is now orders of magnitude slower - therefore I have moved it down the drop-down menu in init.

> Also in the readme better to add a note that it tested for Ubunutu and may work for windows (depends on simplecrf)

I am now working on updating the readme

File: sample-apps/segmentation_spleen_postproc/README.md (hunk 2 above)
Comment (masadcv) [reply sample-apps/segmentation_spleen_postproc/README.md:18]: Thanks, I will keep this version and think about it more later.

File: sample-apps/segmentation_spleen_scribbles/README.md (hunk 7 above)
Comment (masadcv) [reply sample-apps/segmentation_spleen_scribbles/README.md]: Fixed

File: sample-apps/segmentation_spleen_scribbles/main.py (hunk 6 above)
Comment (masadcv) [reply sample-apps/segmentation_spleen_scribbles/main.py:36]: Added description for these methods into README

**Comment by masadcv** [comment]
I have updated the readme with the requested information. This is now ready for merging. I will be here to resolve any remaining issues with this.

**Review by SachidanandAlle — APPROVED** [review]
